ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceRatchet.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.forkchoice;
+
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+/**
+ * Provides a ratchet-like functionality for running fork choice. Callers request that fork choice
+ * be run for a specific slot and this class tracks which slot fork choice has been run for, only
+ * running it again when the slot needs to ratchet up. Thus the slot we run fork choice for only
+ * ever moves up and fork choice is never run for old slots, avoiding unnecessary work.
+ */
+class ForkChoiceRatchet {
```

Comment:
maybe a simple test class to go with this?

Reply:
True - it's covered by the BalanceAttackMigitationForkChoiceStrategyTest because it was split out of there but worth splitting out tests for it as well.